src/ketu/async/sink.clj

```diff
@@ -140,44 +141,35 @@
                :ketu.sink/producer producer
                :ketu.sink/abort-input abort-input}
 
+        threads-done-chan        (async/chan)
+        auto-close-chan          (async/chan)
+
         ;; Channels that are closed when their thread loop is done.
         _ (log/info logger "[sink={}] Start {} worker(s)" sink-name sender-threads-num)
-        sender-threads
-        (for [i (range sender-threads-num)]
-          (let [thread-name (str "ketu-sink-" sink-name "-" i)]
-            (async/thread
-              (try
-                (log/info logger "[sink={} worker={}] Start" sink-name i)
-                (.setName (Thread/currentThread) thread-name)
-                (send-loop state)
-                (catch Exception e
-                  (log/error logger "[sink={} worker={}] Unrecoverable error" sink-name i e))
-                (finally
-                  (log/info logger "[sink={} worker={}] Exit" sink-name i))))))
-
-        ;; A single channel that is closed when all sender-threads are done.
-        sender-threads-done (util/go-drain-all! sender-threads)
-
-        ;; If enabled (default true) close producer when done sending messages from input channel.
-        auto-close-thread
-        (when close-producer?
-          (async/thread
-            (try
-              (log/info logger "[sink={} auto-close] Start" sink-name)
-              (.setName (Thread/currentThread) (str "ketu-sink-" sink-name "-auto-close"))
-
-              (async/<!! sender-threads-done)
-              (log/info logger "[sink={} auto-close] All workers are done" sink-name)
-              (close-producer! state)
-
-              (catch Exception e
-                (log/error logger "[sink={} auto-close] Error" sink-name e)))))
-
-        state (-> state
-                  (assoc :ketu.sink/sender-threads sender-threads
-                         :ketu.sink/sender-threads-done sender-threads-done)
-                  (cond-> auto-close-thread (assoc :ketu.sink/auto-close-thread auto-close-thread)))]
-    state))
+        sender-threads           (mapv #(let [thread-name (str "ketu-sink-" sink-name "-" %)]
+                                          (async/thread
+                                            (try
+                                              (log/info logger "[sink={} worker={}] Start" sink-name %)
+                                              (.setName (Thread/currentThread) thread-name)
+                                              (send-loop state)
+                                              (catch Throwable e
+                                                (log/error logger "[sink={} worker={}] Unrecoverable error" sink-name % e)))
+                                            (log/info logger "[sink={} worker={}] Exit" sink-name %)
```

Comment:
Why? I wish to log `"Unrecoverable error"` only if indeed there was one. Moving it into a `finally` block would always log it, regardless if such an exception occurred or not.

Reply:
1. Sorry I wasn't clear. I mean that the closing logic that comes after the whole try-catch part.
2. I now see that you changed `(catch Exception` to `(catch Throwable`. Catching `Exception` here was intentional, allowing users to implement an UncaughtExceptionHandler and letting them deal with Errors. Once you do that, you need to put the closing code in `finally`, otherwise you might skip it.